### PR TITLE
feat: Unify bounds calculation and make it configurable

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
Related to https://github.com/appium/appium-uiautomator2-server/issues/358#issuecomment-652515990

### CODE REVIEW

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/settings/SimpleBoundsCalculation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633&bid=18949647)

File: app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java [hunk 1]
```diff
@@ -85,37 +87,59 @@ public static String getText(@Nullable AccessibilityNodeInfo nodeInfo, boolean r
...
 
     /**
      * Returns the node's bounds clipped to the size of the display, limited by the MAX_DEPTH
+     * The implementation is borrowed from `getVisibleBounds` method of `UiObject2` class
      *
      * @return null if node is null, else a Rect containing visible bounds
      */
     @SuppressLint("CheckResult")
-    private static Rect getVisibleBounds(@Nullable AccessibilityNodeInfo node, int depth) {
+    private static Rect getBounds(@Nullable AccessibilityNodeInfo node, Rect displayRect, int depth) {
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java:115]: Codacy found an issue: [Avoid reassigning parameters such as 'depth'](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/settings/SimpleBoundsCalculation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633&bid=18949647)

File: app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java (hunk 1 above)
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java:115]: Codacy found an issue: [Avoid reassigning parameters such as 'depth'](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/settings/SimpleBoundsCalculation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633&bid=18949647)

File: app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java [hunk 2]
```diff
@@ -85,37 +87,60 @@ public static String getText(@Nullable AccessibilityNodeInfo nodeInfo, boolean r
...
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java:115]: Codacy found an issue: [Avoid reassigning parameters such as 'depth'](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/settings/SimpleBoundsCalculation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633&bid=18949647)

**Comment by sshock** [comment]
Lgtm; thanks so much for adding this!

Are there appium docs that should also be updated in association with this change?  I.e., the docs that show up here: http://appium.io/docs/en/advanced-concepts/settings/

File: app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java [hunk 3]
```diff
@@ -82,43 +84,62 @@ public static String getText(@Nullable AccessibilityNodeInfo nodeInfo, boolean r
...
+        for (ancestor = node.getParent(); ancestor != null && ++currentDepth < MAX_DEPTH; ancestor = ancestor.getParent()) {
+            if (!ancestor.isScrollable()) {
+                continue;
             }
+            // If this ancestor is scrollable
+            // Trim any portion of the bounds that are hidden by the non-visible portion of our
+            // ancestor
+            Rect ancestorRect = getBounds(ancestor, displayRect, currentDepth);
+            ret.intersect(ancestorRect);
+            break;
```
Comment (dpgraham) [inline app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java:142]: Codacy found an issue: [Avoid using a branching statement as the last in a loop.](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/settings/SimpleBoundsCalculation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633&bid=18949647)

**Comment by dpgraham** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- app/src/main/java/io/appium/uiautomator2/model/settings/SimpleBoundsCalculation.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633&bid=18949647)

**Comment by KazuCocoa** [comment]
@sshock 
Yes, https://github.com/appium/appium/blob/master/docs/en/advanced-concepts/settings.md is.
It is appreciated if you have time to update there :)

**Comment by sshock** [comment]
> It is appreciated if you have time to update there :)

Sure, it's the least I can do.

**Comment by sshock** [comment]
FYI, here is the PR for updating the appium docs: https://github.com/appium/appium/pull/14505

**Pull request merged**